odoo/addons/base/models/ir_access.py

```diff
@@ -0,0 +1,424 @@
+
+    #
+    # ormcache invalidation
+    #
+    def _clear_caches(self):
+        """ Invalidate all the caches related to access rights. """
+        self.env.invalidate_all()
+        self.env.registry.clear_cache()
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        accesses = super().create(vals_list)
+        self._clear_caches()
+        return accesses
+
+    def write(self, values):
+        result = super().write(values)
+        self._clear_caches()
+        return result
+
+    def unlink(self):
+        result = super().unlink()
+        self._clear_caches()
+        return result
+
+    @ormcache('self.env.uid', 'model_name', 'operation', 'tuple(self._get_access_context())')
+    def _get_access_domain(self, model_name: str, operation: str) -> list | None:
+        """ Return the domain that determines on which records of ``model_name``
+        the current user is allowed to perform ``operation``.  The domain comes
+        from the permissions and restrictions that applies to the current user.
+        If no permission exists for the current user, the method returns ``None``.
+        """
+        assert operation in OPERATIONS, "Invalid access operation"
+
+        accesses = self._get_access_records(model_name, operation)
+        if not accesses.group_id:
+            # no group access implies no permission at all
+            return None
+
+        # collect permissions and restrictions
+        permissions = []
+        restrictions = []
+        eval_context = self._eval_context()
+        for access in accesses:
+            domain = safe_eval(access.domain, eval_context) if access.domain else []
+            if access.group_id:
+                permissions.append(domain)
+            else:
+                restrictions.append(domain)
+
+        domain = expression.OR(permissions)
+        if restrictions:
+            domain = expression.AND([domain, *restrictions])
+        return domain
+
+    @ormcache('self.env.uid', 'model_name', 'operation', 'tuple(self._get_access_context())')
+    def _get_restriction_domain(self, model_name: str, operation: str) -> list:
+        """ Return the domain that combines the restrictions to perform
+        ``operation`` on the records of ``model_name``.
+        """
+        assert operation in OPERATIONS, "Invalid access operation"
+
+        domain = [
+            ('model_id', '=', self.env['ir.model']._get_id(model_name)),
+            ('group_id', '=', False),
+            (f'for_{operation}', '=', True),
+            ('active', '=', True),
+        ]
+        accesses = self.sudo().search_fetch(domain, ['group_id', 'domain'], order='id')
+
+        # collect restrictions
+        restrictions = []
+        eval_context = self._eval_context()
+        for access in accesses:
+            domain = safe_eval(access.domain, eval_context) if access.domain else []
+            restrictions.append(domain)
+
+        return expression.AND(restrictions)
```

Comment:
Maybe it could just be this?
```suggestion
        return self.with_user(SUPERUSER_ID)._get_access_domain(model_name, operation)
```